src/Pact/Types/Term.hs

```diff
@@ -417,56 +428,120 @@ data App t = App
 instance Show n => Show (App n) where
   show App{..} = "(" ++ unwords (show _appFun:map show _appArgs) ++ ")"
 
--- TODO: We need a more expressive, safer ADT for this.
-data Module
-  = Module
+
+newtype Governance g = Governance { _gGovernance :: Either KeySetName g }
+  deriving (Eq,Ord,Functor,Foldable,Traversable)
+
+instance Show g => Show (Governance g) where
+  show (Governance (Left k)) = asString' k
+  show (Governance (Right r)) = show r
+
+instance ToJSON g => ToJSON (Governance g) where
+  toJSON (Governance g) = case g of
+    Left ks -> object [ "keyset" .= ks ]
+    Right c -> object [ "capability" .= c ]
+instance FromJSON g => FromJSON (Governance g) where
+  parseJSON = withObject "Governance" $ \o ->
+    Governance <$> (Left <$> o .: "keyset" <|>
+                    Right <$> o .: "capability")
+
+
+data Module g = Module
   { _mName :: !ModuleName
-  , _mKeySet :: !KeySetName
+  , _mGovernance :: !(Governance g)
```

Comment:
We discussed this earlier and I have a better justification for the type parameter: the conversion to `Def Ref` implies all the validation that occurs in `loadModule` so that when we get to granting governance we're not having to throw some illegitimate "this can't be happening" error when the `Maybe (Def Ref)` is `Nothing`. That's what I mean by "unsafe" is there is only convention to guide you w/r/t how much validation to do at various usage sites when there's `Maybe`.

What's more this even applies to my suggestion earlier of a sum type for the cases. When you get to validation you'll still have to handle the possibility of it being a `Name` even if that is impossible in the codebase, because after all that code could change, but `Governance (Def Ref)` cannot (without breaking the code at hand).